File: python/백준/그래프/DFS/24479 알고리즘 수업 - 깊이 우선 탐식 1.py
import sys
sys.setrecursionlimit(10**6)
input = sys.stdin.readline

N, M, R = map(int, input().split())

graph = [[] for _ in range(N+1)]
visited = [0] * (N+1)

for _ in range(M):  # M개 간선 읽기
    a, b = map(int, input().split())
    graph[a].append(b)
    graph[b].append(a)

for i in range(1, N+1):   # 각 인접리스트 오름차순 정렬
    graph[i].sort()

order = 1                 # 방문 순서 카운터

def dfs(start):
    global order
    for neighbor in graph[start]:
        if visited[neighbor] == 0:
            order += 1
            visited[neighbor] = order   # 방문 '순서' 기록
            dfs(neighbor)

visited[R] = order  # 시작 정점 방문 순서 = 1
dfs(R)

for i in range(1, N+1):
    print(visited[i])

File: cpp/백준/그래프/DFS/24479 알고리즘 수업 - 깊이 우선 탐식 1.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int N, M, R;
vector<vector<int>> graph;
vector<int> visited;
int order = 1;

void dfs(int start) {
    for (int neighbor : graph[start]) {
        if (visited[neighbor] == 0) {
            order += 1;
            visited[neighbor] = order;
            dfs(neighbor);
        }
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    cin >> N >> M >> R;
    graph.assign(N + 1, {});
    visited.assign(N + 1, 0);

    for (int i = 0; i < M; ++i) {
        int a, b;
        cin >> a >> b;
        graph[a].push_back(b);
        graph[b].push_back(a);
    }

    for (int i = 1; i <= N; ++i) sort(graph[i].begin(), graph[i].end());

    visited[R] = order;
    dfs(R);

    for (int i = 1; i <= N; ++i) cout << visited[i] << "\n";
    return 0;
}
